e2e/nomostest/watch.go

```diff
@@ -19,95 +19,198 @@ import (
+		watch.timeout = timeout
 	}
+}
 
-	err := watchForObjectInner(nt, gvk, name, namespace, predicates, wait)
-	if err != nil {
-		took := time.Since(startTime)
-		nt.T.Logf("failed after %v watching for predicates to be satisfied", took)
-		switch wait.failureStrategy {
-		case WaitFailureStrategyFatal:
-			nt.T.Fatal(err)
-		case WaitFailureStrategyError:
-			nt.T.Error(err)
-		}
-		return
+// watchErrorOnNotFound configures Watch to error if the object is deleted or
+// not found.
+// This is a private option. You probably want to use WatchForNotFound instead.
+func watchErrorOnNotFound() WatchOption {
+	return func(watch *watchSpec) {
+		watch.errorOnNotFound = true
 	}
-	took := time.Since(startTime)
-	nt.T.Logf("took %v watching for predicates to be satisfied", took)
 }
 
-func watchForObjectInner(nt *NT, gvk schema.GroupVersionKind, name, namespace string, predicates []Predicate, wait waitSpec) error {
-	nt.T.Helper()
+// WatchObjectNotFoundError is returned by WatchObject when the watched object
+// is deleted or not found, if WatchErrorOnDelete is enabled.
+type WatchObjectNotFoundError struct {
+	GroupVersionKind schema.GroupVersionKind
+	Name             string
+	Namespace        string
+}
+
+func (wde *WatchObjectNotFoundError) Error() string {
+	return fmt.Sprintf("expected %s %s/%s to exist and continue to exist, but it was deleted or not found",
+		wde.GroupVersionKind.Kind, wde.Namespace, wde.Name)
+}
+
+// WatchObject watches the specified object util all predicates return nil,
+// or the timeout is reached. Object does not need to exist yet, as long as the
+// resource type exists.
+func WatchObject(nt *NT, gvk schema.GroupVersionKind, name, namespace string, predicates []Predicate, opts ...WatchOption) error {
+	errPrefix := fmt.Sprintf("WatchObject(%s %s/%s)", gvk.Kind, namespace, name)
 
-	listGVK := gvk.GroupVersion().WithKind(gvk.Kind + "List")
-	rObj, err := nt.WatchClient.Scheme().New(listGVK)
+	startTime := time.Now()
+	defer func() {
+		took := time.Since(startTime)
+		nt.T.Logf("%s watched for %v", errPrefix, took)
+	}()
+
+	spec := watchSpec{
+		timeout: nt.DefaultWaitTimeout,
+	}
+	for _, opt := range opts {
+		opt(&spec)
+	}
+
+	cObjList, err := newObjectListForObjectKind(gvk, nt.scheme)
 	if err != nil {
 		return err
 	}
-	cObjList, ok := rObj.(client.ObjectList)
-	if !ok {
-		return fmt.Errorf("%w: got %T, want client.Object", ErrWrongType, rObj)
-	}
 
 	var listOpts []client.ListOption
+	listOpts = append(listOpts, client.MatchingFieldsSelector{
+		Selector: fields.OneTermEqualSelector("metadata.name", name),
+	})
 	if namespace != "" {
 		listOpts = append(listOpts, client.InNamespace(namespace))
 	}
 
-	ctx, cancel := context.WithTimeout(nt.Context, wait.timeout)
+	ctx, cancel := context.WithTimeout(nt.Context, spec.timeout)
 	defer cancel()
+
+	// Use LIST with a single-object selector, instead of GET, because we want
+	// the resource version of the LIST response to start the watch with, so we
```

Comment:
If we use GET, does the watch not capture the resource version change?

Reply:
List and Get have different resource versions. It's the difference between the Object and the ObjectList. How the list versions work, I don't really understand, except to know that they're different things...

Unfortunately the controller-runtime plays fast and loose with this idea (it doesn't pass all the fields through the client.ListOptions to the metav1.ListOptions), so I'm not 100% sure this actually works the same way it does with client-go clients, which use the same strategy for the ListAndWatch that backs the Informer library.

Reply:
ACK. Thanks for the explanation.